org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/core/VMInstallTests.java

```diff
@@ -201,6 +217,114 @@ public void testLibraryResolver4() throws Exception {
 		}
 	}
 
+	private static final Set<String> COMMON_JAVA_PACKAGES = Set.of("java.util", "java.io"); // TODO: TBD
+	private static final String JAVA9_OR_LATER_PACKAGE = "java.lang.module";
+	private static final String JAVA11_OR_LATER_PACKAGE = "java.net.http";
+	private static final String NON_MODULAR_JDK_PACKAGE = "javax.annotation";
+
+	public void testJavaRuntimeQuerySystemPackages_modularJDK() throws CoreException {
+		IVMInstall vm = JavaRuntime.getDefaultVMInstall();
+		String javaVersion = ((IVMInstall2) vm).getJavaVersion();
+		assertTrue("At least a JDK 11 expected", JavaRuntime.compareJavaVersions(vm, JavaCore.VERSION_11) > 0);
+
+		Set<String> defaultVersionPackages = JavaRuntime.getProvidedVMPackages(vm, null);
+		assertTrue(defaultVersionPackages.containsAll(COMMON_JAVA_PACKAGES));
+		assertTrue(defaultVersionPackages.contains(JAVA9_OR_LATER_PACKAGE));
+		assertTrue(defaultVersionPackages.contains(JAVA11_OR_LATER_PACKAGE));
+		assertFalse(defaultVersionPackages.contains(NON_MODULAR_JDK_PACKAGE));
+
+		Set<String> java9Packages = JavaRuntime.getProvidedVMPackages(vm, "9");
+		assertTrue(java9Packages.containsAll(COMMON_JAVA_PACKAGES));
+		assertTrue(defaultVersionPackages.contains(JAVA9_OR_LATER_PACKAGE));
+		assertFalse(java9Packages.contains(JAVA11_OR_LATER_PACKAGE));
+		assertFalse(java9Packages.contains(NON_MODULAR_JDK_PACKAGE));
+
+		// Test that a null release is equals to the actual version of the VM
+		assertEquals(defaultVersionPackages, JavaRuntime.getProvidedVMPackages(vm, javaVersion));
+
+		// Test failure if non-modular VM is asked for packages of Java<=1.8
+		CoreException e1 = assertThrows(CoreException.class, () -> JavaRuntime.getProvidedVMPackages(vm, "1.8"));
+		assertEquals("Cannot query a modular VM (JavaSE-9 or higher) for packages of JavaSE-1.8 and lower", e1.getMessage());
+		JavaRuntime.compareJavaVersions(vm, javaVersion);
+		String nextJavaVersion = Integer.toString(Integer.parseInt(javaVersion.substring(0, javaVersion.indexOf('.'))) + 1);
+		CoreException e2 = assertThrows(CoreException.class, () -> JavaRuntime.getProvidedVMPackages(vm, nextJavaVersion));
+		assertEquals("release " + nextJavaVersion + " is not found in the system", e2.getMessage());
+	}
+
+	public void testJavaRuntimeQuerySystemPackages_nonModularJDK() throws Exception {
+		try (AutoCloseableSupplier<IVMInstall> nonModularVM = searchFirstNonModularVM()) {
+			IVMInstall vm = nonModularVM.get();
+
+			Set<String> packages = JavaRuntime.getProvidedVMPackages(vm, null);
+
+			LinkedHashSet<String> missing = new LinkedHashSet<>(COMMON_JAVA_PACKAGES);
+			missing.removeAll(packages);
+			assertTrue("Not all packages found in " + vm.getInstallLocation() + ", missing " + missing, missing.isEmpty());
```

Comment:
Changed assertions to provide meaningful fail message if some package is missing or not present